tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -50,82 +50,93 @@ astDeclarationsForLanguage language filePath = do
+constructorForSumChoice :: String -> Name -> TreeSitter.Deserialize.Type -> Q Con
+constructorForSumChoice str typeParameterName (MkType (DatatypeName n) named) = normalC (toName named (n ++ str)) [child]
+  where child = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (appT (conT (toName named n)) (varT typeParameterName))
 
 -- | Build Q Constructor for product types (nodes with fields)
-ctorForProductType :: String -> Maybe Children -> [(String, Field)] -> Q Con
-ctorForProductType constructorName children fields = recC (toName Named constructorName) lists where
-  lists = fieldList ++ childList
-  fieldList = fmap (uncurry toVarBangType) fields
-  childList = toList $ fmap toVarBangTypeChild children
-  toVarBangType name (MkField required fieldTypes mult) =
-    let fieldName = mkName . addTickIfNecessary . removeUnderscore $ name
-        strictness = TH.bang noSourceUnpackedness noSourceStrictness
-        ftypes = fieldTypesToNestedEither fieldTypes
-        fieldContents = case (required, mult) of
-          (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
-          (Required, Single) -> ftypes
-          (Optional, Multiple) -> appT (conT ''[]) ftypes
-          (Optional, Single) -> appT (conT ''Maybe) ftypes
-    in TH.varBangType fieldName (TH.bangType strictness fieldContents)
-  toVarBangTypeChild (MkChildren field) = toVarBangType "extra_children" field
+ctorForProductType :: String -> Name -> Maybe Children -> [(String, Field)] -> Q Con
+ctorForProductType constructorName typeParameterName children fields = ctorForTypes constructorName lists where
+  lists = annotation : fieldList ++ childList
+  annotation = ("ann", varT typeParameterName)
+  fieldList = map (fmap toType) fields
+  childList = toList $ fmap toTypeChild children
+  toType (MkField required fieldTypes mult) =
+    let ftypes = fieldTypesToNestedEither fieldTypes typeParameterName
+    in case (required, mult) of
+      (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
+      (Required, Single) -> ftypes
+      (Optional, Multiple) -> appT (conT ''[]) ftypes
+      (Optional, Single) -> appT (conT ''Maybe) ftypes
+  toTypeChild (MkChildren field) = ("extra_children", toType field)
 
 
 -- | Build Q Constructor for leaf types (nodes with no fields or subtypes)
-ctorForLeafType :: Named -> DatatypeName -> Q Con
-ctorForLeafType Anonymous (DatatypeName name) = normalC (toName Anonymous name) []
-ctorForLeafType Named (DatatypeName name) = recC (toName Named name) [leafBytes] where
-  leafBytes = TH.varBangType (mkName "bytes") textValue
-  textValue = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT ''Text)
+ctorForLeafType :: Named -> DatatypeName -> Name -> Q Con
+ctorForLeafType Anonymous (DatatypeName name) typeParameterName = normalC (toName Anonymous name) [TH.bangType strictness (varT typeParameterName)]
+  where strictness = TH.bang noSourceUnpackedness noSourceStrictness
```

Comment:
We use this same definition in a couple of different places. Would it be worth extracting it to the top level?

Reply:
Done.